docs/spec.md

```diff
@@ -0,0 +1,406 @@
+
+- AI models can securely access Storacha's decentralized storage
+- Applications can store and retrieve data without custom integration code
+- The MCP standard ensures compatibility with a growing ecosystem of clients
+
+## Storacha MCP Server Architecture
+
+### Basic Flow
+
+```mermaid
+flowchart LR
+    A[AI Application] -->|HTTP Request| B[Storacha MCP Server]
+    B -->|Upload| C[w3up-client]
+    B -->|Retrieve| E[HTTP Gateway]
+    C -->|Store| D[(Decentralized Storage)]
+    D -->|Content-addressed Data| E
+    C -->|Response| B
+    E -->|Response| B
+    B -->|HTTP Response| A
+    
+    classDef client fill:#d4f1f9,stroke:#0078d7,stroke-width:2px
+    classDef server fill:#ffe6cc,stroke:#d79b00,stroke-width:2px
+    classDef storage fill:#d5e8d4,stroke:#82b366,stroke-width:2px
+    classDef gateway fill:#fff2cc,stroke:#d6b656,stroke-width:2px
+    
+    class A client
+    class B server
+    class C,D storage
+    class E gateway
+```
+
+### Core Functionality
+
+The Storacha MCP server will support two primary operations in this first iteration:
+
+1. **Upload**: Store files on Storacha's decentralized storage using the Storacha client
+2. **Retrieve**: Get files from Storacha by their Content ID (CID) through the Storacha HTTP trustless Gateway
+3. **Authentication & Authorization**: Optional authentication via bearer tokens
+
+#### Upload Operation
+
+- **Multiple File Support**: The upload operation can handle one or more files in a single request
+- **Universal File Types**: Supports any file type, which is automatically converted to a blob by the Storacha client
+- **Content Addressing**: Returns Content IDs (CIDs) for uploaded files, enabling permanent, verifiable references
+- **Space Delegation**: Files are uploaded to the space authorized by the provided delegation in the server configuration
+- **Flexible Space Delegation**: Can accept delegation proof in request headers or body, with runtime delegations taking priority over server configuration
+
+#### Retrieve Operation
+
+- **CID-Based Access**: Files are retrieved using their unique Content IDs (CIDs)
+- **Gateway Access**: Retrieval happens through the Storacha HTTP trustless gateway (https://storacha.link) rather than the client instance, but the users can provide a custom gateway in the server configuration if needed
+- **Format Conversion**: Files are returned in a format compatible with the MCP specification (base64 data URIs)
+- **Metadata Preservation**: Content types and other metadata are preserved and returned with the file
+- **Public Access**: No authentication required for retrieval, as CIDs are self-authenticating
+
+#### Authentication & Authorization
+
+- **Bearer Token Authentication**: Optional API key authentication using bearer tokens to control access to the upload operation
+- **Delegation Priority**: Runtime delegations provided in requests take precedence over server pre defined delegation configuration
```

Comment:
👌 nice touch

Reply:
I'd perhaps mention UCAN here.

Reply:
Thanks. Yes, good call. I will add that too.

Reply:
Done.